--- sh/init.sh.Linux.in 2015-02-05 04:03:26.000000000 +0000
+++ sh/init.sh.Linux.in 2015-02-07 20:40:15.559562712 +0000
@@ -46,8 +46,8 @@
 		fi
 		mkdir /run
 	else
-		eerror "The /run directory does not exist. Unable to continue."
-		return 1
+		eerror "The /run directory does not exist - cannot continue"
+		exit 1
 	fi
 fi
 
@@ -55,20 +55,19 @@
 	rm -rf /run/*
 elif ! mountinfo -q /run; then
 	ebegin "Mounting /run"
-	rc=0
 	if ! fstabinfo --mount /run; then
-		mount -t tmpfs -o mode=0755,nodev,size=10% tmpfs /run
-		rc=$?
+		mount -t tmpfs -o mode=1775,gid=500,size=8M,nosuid,noexec tmpfs /run
 	fi
+	rc=$?
+	eend $rc
 	if [ $rc != 0 ]; then
-		eerror "Unable to mount tmpfs on /run."
-		eerror "Can't continue."
+		eerror "Unable to mount tmpfs on /run - cannot continue"
 		exit 1
 	fi
 fi
 
-checkpath -d $RC_SVCDIR
-checkpath -d -m 0775 -o root:uucp /run/lock
+checkpath -d -m 0700 -o root:root $RC_SVCDIR >/dev/null 2>&1
+checkpath -d -m 0755 -o root:root /run/lock >/dev/null 2>&1
 
 # Try to mount xenfs as early as possible, otherwise rc_sys() will always
 # return RC_SYS_XENU and will think that we are in a domU while it's not.
