--- smtpd/ssl.c 2015-06-30 08:13:34.000000000 +0000
+++ smtpd/ssl.c 2015-07-03 18:32:33.534060771 +0000
@@ -278,7 +278,11 @@
 	SSL_CTX	       *ctx;
 	size_t		pkinamelen = 0;
 
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
 	ctx = SSL_CTX_new(SSLv23_method());
+#else
+	ctx = SSL_CTX_new(TLS_method());
+#endif
 	if (ctx == NULL) {
 		ssl_error("ssl_ctx_create");
 		fatal("ssl_ctx_create: could not create SSL context");
@@ -286,10 +290,22 @@
 
 	SSL_CTX_set_session_cache_mode(ctx, SSL_SESS_CACHE_OFF);
 	SSL_CTX_set_timeout(ctx, SSL_SESSION_TIMEOUT);
-	SSL_CTX_set_options(ctx,
-	    SSL_OP_ALL | SSL_OP_NO_SSLv2 | SSL_OP_NO_SSLv3 | SSL_OP_NO_TICKET);
-	SSL_CTX_set_options(ctx,
-	    SSL_OP_NO_SESSION_RESUMPTION_ON_RENEGOTIATION);
+
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
+	SSL_CTX_set_options(ctx, SSL_OP_NO_SSLv2);
+	SSL_CTX_set_options(ctx, SSL_OP_NO_SSLv3);
+#endif
+
+	SSL_CTX_set_options(ctx, SSL_OP_CIPHER_SERVER_PREFERENCE);
+	SSL_CTX_set_options(ctx, SSL_OP_NO_COMPRESSION);
+	SSL_CTX_set_options(ctx, SSL_OP_NO_SESSION_RESUMPTION_ON_RENEGOTIATION);
+	SSL_CTX_set_options(ctx, SSL_OP_NO_TICKET);
+	SSL_CTX_set_options(ctx, SSL_OP_SINGLE_DH_USE);
+	SSL_CTX_set_options(ctx, SSL_OP_SINGLE_ECDH_USE);
+
+#ifdef SSL_OP_DONT_INSERT_EMPTY_FRAGMENTS
+	SSL_CTX_clear_options(ctx, SSL_OP_DONT_INSERT_EMPTY_FRAGMENTS);
+#endif
 
 	if (ciphers == NULL)
 		ciphers = SSL_CIPHERS;
